### PR TITLE
Add analyzers from FSharp.Formatting PR

### PR DESCRIPTION
**Pull request by nojaf**
In https://github.com/fsprojects/FSharp.Formatting/pull/906, @thorium made some good suggestions to improve the performance of the code.

Some of these checks are interesting to turn into an analyzer I think.
I will try and add some analyzers in this PR.

- [X] HeadConsEmptyListPatternAnalyzer
- [x] `list = []` analyzer (should be `List.empty`)
- [x] Propose `[<return: Struct>]` on (partial) active pattern.
- [x] `List.map` piped to `List.map` call. 
- [x] `a = null`, should be `isNull a`
- [x] `a <> null` should be `not (isNull a)`
- [x] Add `[<Struct>]` to DU when it makes sense. (For example each DU case does not have any fields)

### CODE REVIEW

**Comment by Thorium** [comment]
I have more suggestions if you run out :-D

- [ ] .ContainsKey key followed by .[key] should be TryGetValue
- [ ] ConcurrentDictionaries shouldn't be called with Add or .[x] but rather via GetOrAdd or AddOrUpdate
- [ ] Seq.length > 0 and (x:seq).Length > 0 should be replaced with Seq.isEmpty
- [ ] If you have large list/array/seq of items and you do .Contains in a loop, consider option to change your list to Set
- [ ] Don't do new Regex.Match in a loop: rather let myRegex = Regex("...") and then loop your myRegex.Match
- [ ] Also always try to close your regex, to make it “StartsWith” or “EndsWith” and not “Contains”
- [ ] Don't do DateTime.UtcNow.Date, it points to a random date that is not what you wanted.
- [ ] If you ever do TransactionScope remember to set TransactionScopeAsyncFlowOption to 1, as modern software is asynchronous.
- [ ] Besides of this I did 10 years ago a list that I hoped would be "functional refactorings" in a "what is ReSharper to C#" spirit: https://www.slideshare.net/thorium/possible-fsharp-refactorings-could-be

**Comment by nojaf** [comment]
> I have more suggestions if you run out :-D

Thanks, that is good input. Realistically speaking I'll probably not get to those.
My time for this is limited.

**Comment by Numpsy** [comment]
I'm sure there are plenty of possible ideas for analyzers about (be it borrowing more rules from FSharpLint, or looking at things like NetAnalyzers and Roslynator, or more F# specific stuff) - is it worth creating a more general list of suggestions somewhere to track ideas?

**Comment by Thorium** [comment]
Something that ionide-analysers could in theory do, and not so much of others (just because they are not F# IDE-plugins), would be these 2 way conversions of "change A to B" and "change B to A" and leave final decision for user to decide. Because that would help refactorings.

**Comment by nojaf** [comment]
@dawedawe feel free to already start nitpicking if you like.

File: docs/performance/008.md
Comment (dawedawe) [inline docs/performance/008.md:23]:
```suggestion
let b = List.isEmpty a // will check if there is no tail which is faster than the equality check.
```

File: docs/performance/010.md
Comment (dawedawe) [inline docs/performance/010.md:12]:
```suggestion
Subsequent collection functions like `List.filter` and `List.map` can sometimes be combined into `List.choose`.
```

**Review by dawedawe — COMMENTED** [review]
Very solid work!

File: docs/performance/010.md
Comment (Thorium) [reply docs/performance/010.md]: Besides of List.choose, there are a few other map+reduce functions that can combine other list-operations to avoid iterating lists multiple times through, e.g. instead of doing list-operations and then List.concat in the end, it can sometimes be replaced with List.collect. And sometimes List.map + List.sum can be replaced with List.sumBy.

File: tests/Ionide.Analyzers.Tests/Suggestion/StructDiscriminatedUnionAnalyzerTests.fs
Comment (Thorium) [inline tests/Ionide.Analyzers.Tests/Suggestion/StructDiscriminatedUnionAnalyzerTests.fs:66]: There is a catch here: Tuple is class and not primitive. ValueTuple is a struct.

Comment (nojaf) [reply tests/Ionide.Analyzers.Tests/Suggestion/StructDiscriminatedUnionAnalyzerTests.fs]: Thanks, I didn't think about that.

File: docs/performance/010.md
Comment (nojaf) [reply docs/performance/010.md]: Thanks, good input. I'm going to consider this out of scope for this PR, feel free to create an extension of this PR once it gets merged.

**Comment by Numpsy** [comment]
A question about the ```Add [<Struct>] to DU when it makes sense.``` suggestion/fix:

Say I have a DU in one of  these forms:
```fsharp
type SomeType =
| Success
| SomeError of string
| SomeOtherError of string

type SomeType =
| Success
| SomeError of error: string
| SomeOtherError of error: string
```
I get the 'make it a struct' suggestion and quick fix:
image
But if I action the quick fix, I get:
image

Maybe making sure the cases have unique names is a good suggestion anyway, but fwiw - should the quick fix make a change that then errors?

**Comment by nojaf** [comment]
> A question about the Add [<Struct>] to DU when it makes sense. suggestion/fix:

I've noticed the issue as well. Ideally, the quick fix should address this. I'm inclined to accept the quick fix as it stands. While it's not ideal, anyone interested can further refine it with subsequent PRs. I feel this option might be the better choice compared to turning off the code fix and nobody ever touches it again.

This PR's work is supported by the Amplifying F# Open Collective (more details to be shared publicly later this week). We allocated one working day for this task, and we've now surpassed that limit. I'll be shifting my focus to other work that supports my livelihood.

**Review by dawedawe — APPROVED** [review]
Great work. You really squeezed the funding to the max. Thanks!
The known limitations should be handled in another round.

**Pull request merged**
